controllers/authpolicy_envoysecuritypolicy_controller.go

```diff
@@ -0,0 +1,190 @@
+package controllers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	egv1alpha1 "github.com/envoyproxy/gateway/api/v1alpha1"
+	"github.com/go-logr/logr"
+	kuadrantv1beta2 "github.com/kuadrant/kuadrant-operator/api/v1beta2"
+	kuadrantenvoygateway "github.com/kuadrant/kuadrant-operator/pkg/envoygateway"
+	"github.com/kuadrant/kuadrant-operator/pkg/kuadranttools"
+	kuadrantgatewayapi "github.com/kuadrant/kuadrant-operator/pkg/library/gatewayapi"
+	"github.com/kuadrant/kuadrant-operator/pkg/library/kuadrant"
+	"github.com/kuadrant/kuadrant-operator/pkg/library/mappers"
+	"github.com/kuadrant/kuadrant-operator/pkg/library/reconcilers"
+	"github.com/kuadrant/kuadrant-operator/pkg/library/utils"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	gatewayapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+	gatewayapiv1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+)
+
+// AuthPolicyEnvoySecurityPolicyReconciler reconciles SecurityPolicy objects for auth
+type AuthPolicyEnvoySecurityPolicyReconciler struct {
+	*reconcilers.BaseReconciler
+}
+
+//+kubebuilder:rbac:groups=gateway.envoyproxy.io,resources=securitypolicies,verbs=get;list;watch;create;update;patch;delete
+
+func (r *AuthPolicyEnvoySecurityPolicyReconciler) Reconcile(eventCtx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := r.Logger().WithValues("Gateway", req.NamespacedName)
+	logger.Info("Reconciling auth SecurityPolicy")
+	ctx := logr.NewContext(eventCtx, logger)
+
+	gw := &gatewayapiv1.Gateway{}
+	if err := r.Client().Get(ctx, req.NamespacedName, gw); err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("no gateway found")
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, "failed to get gateway")
+		return ctrl.Result{}, err
+	}
+
+	if logger.V(1).Enabled() {
+		jsonData, err := json.MarshalIndent(gw, "", "  ")
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		logger.V(1).Info(string(jsonData))
+	}
+
+	if !kuadrant.IsKuadrantManaged(gw) {
+		return ctrl.Result{}, nil
+	}
+
+	topology, err := kuadranttools.TopologyFromGateway(ctx, r.Client(), gw, &kuadrantv1beta2.AuthPolicy{})
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	kuadrantNamespace, err := kuadrant.GetKuadrantNamespace(gw)
+	if err != nil {
+		logger.Error(err, "failed to get kuadrant namespace")
+		return ctrl.Result{}, err
+	}
+
+	// reconcile security policies for gateways
+	for _, gwNode := range topology.Gateways() {
```

Comment:
according the reconciler setup, there is only one gw. No need to iterate. Is it needed?

Reply:
I iterated to handle case of 0 or 1 gateway instead of checking length > 0 before taking first index but I could do that instead if you prefer. I thought this way looked slightly cleaner

Reply:
It should always be 1, but this is good enough for now.

Reply:
Right yep I see, only if the gateway is not programmed will it be 0